In [7]:
import pandas as pd 
pd.set_option("max_colwidth", 150)

f = "https://github.com/cs6220/cs6220.spring2019/raw/master/data/Online%20Retail.xlsx"
df = pd.read_excel(f)
basket = (df[df["Country"] == "United Kingdom"].groupby(["InvoiceNo", "Description"])["Quantity"].sum().unstack().reset_index().fillna(0).set_index("InvoiceNo")) # transform transactions into baskets of items
basket_sets = basket.map(lambda x: True if x >=1 else False) # convert counts to booleans

## 1.1 Frequent Itemset Generation

**1. What are the top 5 1-itemsets with the highest support?**

In [10]:
basket_sets.head()

Description  20713   4 PURPLE FLOCK DINNER CANDLES  \
InvoiceNo                                            
536365       False                           False   
536366       False                           False   
536367       False                           False   
536368       False                           False   
536369       False                           False   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536365                                False               False   
536366                                False               False   
536367                                False               False   
536368                                False               False   
536369                                False               False   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
InvoiceNo                                                             
536365                              False                     False   
536366                              False                     False   
536367                              False                     False   
536368                              False                     False   
536369                              False                     False   

Description   OVAL WALL MIRROR DIAMANTE    RED SPOT GIFT BAG LARGE  \
InvoiceNo                                                            
536365                             False                     False   
536366                             False                     False   
536367                             False                     False   
536368                             False                     False   
536369                             False                     False   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  ...  \
InvoiceNo                                                               ...   
536365                                  False                    False  ...   
536366                                  False                    False  ...   
536367                                  False                    False  ...   
536368                                  False                    False  ...   
536369                                  False                    False  ...   

Description  wrongly coded 20713  wrongly coded 23343  wrongly coded-23343  \
InvoiceNo                                                                    
536365                     False                False                False   
536366                     False                False                False   
536367                     False                False                False   
536368                     False                False                False   
536369                     False                False                False   

Description  wrongly marked  wrongly marked 23343  \
InvoiceNo                                           
536365                False                 False   
536366                False                 False   
536367                False                 False   
536368                False                 False   
536369                False                 False   

Description  wrongly marked carton 22804  wrongly marked. 23343 in box  \
InvoiceNo                                                                
536365                             False                         False   
536366                             False                         False   
536367                             False                         False   
536368                             False                         False   
536369                             False                         False   

Description  wrongly sold (22719) barcode  wrongly sold as sets  \
InvoiceNo                                                         
536365                              

In [13]:
from mlxtend.frequent_patterns import apriori

In [35]:
freq_items=apriori(basket_sets, min_support=0.02, use_colnames=True)
freq_items.sort_values(by='support',ascending=False)

support                                       itemsets
188  0.098276           (WHITE HANGING HEART T-LIGHT HOLDER)
75   0.087931                      (JUMBO BAG RED RETROSPOT)
148  0.076452                     (REGENCY CAKESTAND 3 TIER)
121  0.072323                                (PARTY BUNTING)
99   0.063158                      (LUNCH BAG RED RETROSPOT)
..        ...                                            ...
170  0.020327                 (SMALL HEART MEASURING SPOONS)
181  0.020327                       (TOY TIDY PINK POLKADOT)
21   0.020191                (CHOCOLATE THIS WAY METAL SIGN)
229  0.020145  (LUNCH BAG SUKI DESIGN , LUNCH BAG CARS BLUE)
114  0.020009               (PACK OF 60 SPACEBOY CAKE CASES)

[242 rows x 2 columns]

In [37]:
freq_1itemset = freq_items[freq_items['itemsets'].apply(lambda x: len(x) == 1)]

In [41]:
top5_freq_1itemset = freq_1itemset.sort_values(by='support',ascending=False).head()
top5_freq_1itemset

support                              itemsets
188  0.098276  (WHITE HANGING HEART T-LIGHT HOLDER)
75   0.087931             (JUMBO BAG RED RETROSPOT)
148  0.076452            (REGENCY CAKESTAND 3 TIER)
121  0.072323                       (PARTY BUNTING)
99   0.063158             (LUNCH BAG RED RETROSPOT)

Hence, top 5 itemset with highest support are: 
1. WHITE HANGING HEART T-LIGHT HOLDER
2. JUMBO BAG RED RETROSPO
3. REGENCY CAKESTAND 3 TIER
4. PARTY BUNTING
5. LUNCH BAG RED RETROSPOT

**2. What are the top 5 2-itemsets with the highest support?**

In [63]:
freq_2itemsets = freq_items[freq_items['itemsets'].apply(lambda x: len(x) == 2)]

In [65]:
top5_freq_2itemsets = freq_2itemsets.sort_values(by='support',ascending=False).head()
top5_freq_2itemsets

support  \
211  0.035617   
207  0.031806   
218  0.031670   
217  0.029809   
223  0.027541   

                                                                itemsets  
211                   (JUMBO BAG PINK POLKADOT, JUMBO BAG RED RETROSPOT)  
207  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )  
218                    (JUMBO STORAGE BAG SUKI, JUMBO BAG RED RETROSPOT)  
217         (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG RED RETROSPOT)  
223                   (LUNCH BAG RED RETROSPOT, LUNCH BAG  BLACK SKULL.)

**3. What is the highest support value for the 1-itemsets?**

From top5_freq_1itemset above, we can see the highest support value for the 1-itemsets is **0.098276**	

**4. What is the highest support value for the 2-itemsets?**

From top5_freq_2itemsets above, we can see the highest support value for the 2-itemsets is **0.035617**	

## 1.2 Association Rule Generation

1. What are the top 5 association rules?

In [43]:
from mlxtend.frequent_patterns import association_rules

In [70]:
rules = association_rules(freq_items, metric="confidence", min_threshold=0.7).sort_values(by='confidence',ascending=False).head()
rules

antecedents  \
8  (ROSES REGENCY TEACUP AND SAUCER , PINK REGENCY TEACUP AND SAUCER)   
7   (GREEN REGENCY TEACUP AND SAUCER, PINK REGENCY TEACUP AND SAUCER)   
2                                    (PINK REGENCY TEACUP AND SAUCER)   
5                                    (PINK REGENCY TEACUP AND SAUCER)   
3                                   (GREEN REGENCY TEACUP AND SAUCER)   

                          consequents  antecedent support  consequent support  \
8   (GREEN REGENCY TEACUP AND SAUCER)            0.024773            0.042377   
7  (ROSES REGENCY TEACUP AND SAUCER )            0.026180            0.043421   
2   (GREEN REGENCY TEACUP AND SAUCER)            0.031897            0.042377   
5  (ROSES REGENCY TEACUP AND SAUCER )            0.031897            0.043421   
3  (ROSES REGENCY TEACUP AND SAUCER )            0.042377            0.043421   

    support  confidence       lift  representativity  leverage  conviction  \
8  0.022368    0.902930  21.306837               1.0  0.021319    9.865319   
7  0.022368    0.854419  19.677538               1.0  0.021232    6.570786   
2  0.026180    0.820768  19.368019               1.0  0.024828    5.342926   
5  0.024773    0.776671  17.886978               1.0  0.023388    4.283280   
3  0.031806    0.750535  17.285056               1.0  0.029966    3.834527   

   zhangs_metric   jaccard  certainty  kulczynski  
8       0.977277  0.499493   0.898635    0.715384  
7       0.974698  0.473583   0.847811    0.684785  
2       0.979615  0.544340   0.812837    0.719271  
5       0.975199  0.490126   0.766534    0.673602  
3       0.983839  0.589076   0.739212    0.741516

2. What items make up one of the top association rules? Search online for the items (or at least items with the same name). Do you think they are likely to be bought together?

REGENCY TEACUP AND SAUCER of different colors.  Yes, they're very likely to be bought together since people always need more than just 1 set of teacup and saucer.

## 2.1 Association Rule Mining

**1. Transform the raw dataset into a format appropriate for association rule mining by dropping all continuous columns and one-hot encoding the remaining columns. The values for each resulting column should be binary, represented by a 1 or 0.**

**2. Use confidence for the rule interestingness (metric="confidence") and gen- erate rules up to a depth of 3 (max len=3). Generate rules and flag at least 5 examples that you find interesting. Comment on your findings and share a few notes on what you learned from the rules you are highlighting. Note that you can choose your support and confidence thresholds yourself in this analysis.**

**3. Use lift for the rule interestingness (metric="lift") and generate rules up to a depth of 3 (max len=3). Generate rules and flag at least 5 examples that you find interesting. Comment on your findings and share a few notes on what you learned from the rules you are highlighting. Note that you can choose your support and confidence thresholds yourself in this analysis.**

**4. Compare the top rules using the two interestingness measures for the same levels of support (use at least two different levels of support) and comment on your findings.**

In [12]:
basket

Description  20713   4 PURPLE FLOCK DINNER CANDLES  \
InvoiceNo                                            
536365         0.0                             0.0   
536366         0.0                             0.0   
536367         0.0                             0.0   
536368         0.0                             0.0   
536369         0.0                             0.0   
...            ...                             ...   
C581484        0.0                             0.0   
C581490        0.0                             0.0   
C581499        0.0                             0.0   
C581568        0.0                             0.0   
C581569        0.0                             0.0   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536365                                  0.0                 0.0   
536366                                  0.0                 0.0   
536367                                  0.0                 0.0   
536368                                  0.0                 0.0   
536369                                  0.0                 0.0   
...                                     ...                 ...   
C581484                                 0.0                 0.0   
C581490                                 0.0                 0.0   
C581499                                 0.0                 0.0   
C581568                                 0.0                 0.0   
C581569                                 0.0                 0.0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
InvoiceNo                                                             
536365                                0.0                       0.0   
536366                                0.0                       0.0   
536367                                0.0                       0.0   
536368                                0.0                       0.0   
536369                                0.0                       0.0   
...                                   ...                       ...   
C581484                               0.0                       0.0   
C581490                               0.0                       0.0   
C581499                               0.0                       0.0   
C581568                               0.0                       0.0   
C581569                               0.0                       0.0   

Description   OVAL WALL MIRROR DIAMANTE    RED SPOT GIFT BAG LARGE  \
InvoiceNo                                                            
536365                               0.0                       0.0   
536366                               0.0                       0.0   
536367                               0.0                       0.0   
536368                               0.0                       0.0   
536369                               0.0                       0.0   
...                                  ...                       ...   
C581484                              0.0                       0.0   
C581490                              0.0                       0.0   
C581499                              0.0                       0.0   
C581568                              0.0                       0.0   
C581569                              0.0                       0.0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  ...  \
InvoiceNo                                                               ...   
536365                                    0.0                      0.0  ...   
536366                                    0.0                      0.0  ...   
536367                                    0.0                      0.0  ...   
536368                                    0.0                      0.0  ...   
536369                                    0.0                      0.0  ...   
...                                  

In [13]:
# basket.dtypes
non_numeric_columns = basket.select_dtypes(exclude=['number'])
print("not numeric ",non_numeric_columns)
# basket['536365']

not numeric  Empty DataFrame
Columns: []
Index: [536365, 536366, 536367, 536368, 536369, 536371, 536372, 536373, 536374, 536375, 536376, 536377, 536378, 536380, 536381, 536382, 536384, 536385, 536386, 536387, 536388, 536390, 536392, 536393, 536394, 536395, 536396, 536397, 536398, 536399, 536400, 536401, 536402, 536404, 536405, 536406, 536407, 536408, 536409, 536412, 536415, 536416, 536420, 536423, 536425, 536437, 536446, 536460, 536463, 536464, 536466, 536477, 536488, 536500, 536502, 536508, 536514, 536520, 536521, 536522, 536523, 536524, 536525, 536526, 536528, 536529, 536530, 536531, 536533, 536534, 536535, 536536, 536537, 536538, 536539, 536542, 536544, 536551, 536555, 536556, 536557, 536558, 536559, 536560, 536561, 536562, 536563, 536564, 536565, 536566, 536567, 536568, 536569, 536570, 536571, 536572, 536573, 536574, 536575, 536576, ...]

[22040 rows x 0 columns]
